### PR TITLE
[tutorial] Relay pass infra tutorial

### PR DESCRIPTION
**Pull request by zhiics**
Add a tutorial for the pass infra.

### CODE REVIEW

**Comment by billzorn** [comment]
Is there a tutorial for adding a new custom pass, completely in Python?

I think it would be useful to have some general, easy to find documentation on "working with the Relay IR / AST in Python." There are applications beyond just optimizing programs - I found while writing a tool to convert Relay to FPCore that working with the IR in Python was actually very convenient, but I had to have a Relay expert on hand to explain how to do it.

**Comment by zhiics** [comment]
@billzorn Sorry for the late response. I've been quite busy in the past several weeks. I will try to find some cycles next week to finish a version for this tutorial so that we can start reviewing. 

I've covered an example in this tutorial about writing a customized pass in Python. You can probably take a look. Any feedback is highly appreciated.

**Comment by zhiics** [comment]
@weberlo @slyubomirsky @billzorn @tqchen @wweic @yongwww @jroesch please review when you have time. Thanks.

File: tutorials/dev/relay_pass_infra.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+# currently doesn't work well on let bindings. Therefore, we would not be able
+# to fuse operators that were fusable if ToANormalForm is applied before
+# fusion, as this pass generates let bindings for each expression to
+# canonicalize a Relay program.
+#
+# Relay, hence, provides `Sequential`_ to alleviate developers from handling
+# these issues explicitly by specifying the required passes of each pass and
+# packing them as a whole to execute. For example, the same passes can now be
+# applied using the sequential style as the following. `Sequential`_ is
+# similiar to `torch.nn.sequential <https://pytorch.org/docs/stable/nn.html#torch.nn.Sequential>`_
```
Comment (yzhliu) [inline tutorials/dev/relay_pass_infra.py]: similar
```suggestion
# similar to `torch.nn.sequential <https://pytorch.org/docs/stable/nn.html#torch.nn.Sequential>`_
```

File: tutorials/dev/relay_pass_infra.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+print(mod4)
+
+seq1 = relay.transform.Sequential([relay.transform.AlterOpLayout()])
+with relay.build_config(opt_level=3):
+    with tvm.target.create("llvm"):
+        mod5 = seq1(mod)
+print(mod5)
+
+##############################################################################
+# Implement a Pass Using Python Syntax Sugar
```
Comment (MarisaKirisame) [inline tutorials/dev/relay_pass_infra.py]:
```suggestion
# Implement a Pass Using Python Decorator
```

File: tutorials/dev/relay_pass_infra.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+# a signle primitive op.
+print(mod)
+
+###############################################################################
+# Use `Sequential`_ to Apply a Sequence of Passes
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Applying passes as above is actually tedious and it may require users to have
+# better understanding about the dependencies between them. For example, fusion
+# currently doesn't work well on let bindings. Therefore, we would not be able
+# to fuse operators that were fusable if ToANormalForm is applied before
```
Comment (yongwww) [inline tutorials/dev/relay_pass_infra.py]: add link (https://docs.tvm.ai/api/python/relay/transform.html?highlight=toanormalform#tvm.relay.transform.ToANormalForm) for `ToANormalForm` here?

Comment (yzhliu) [reply tutorials/dev/relay_pass_infra.py]: @zhiics update?

Comment (zhiics) [reply tutorials/dev/relay_pass_infra.py]: @yzhliu Sorry. I forgot. Just updated.

**Comment by yzhliu** [comment]
Thanks @zhiics @billzorn @MarisaKirisame @yongwww

**Pull request merged**